sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -655,6 +655,30 @@ object TypeCoercion {
       case NaNvl(l, r) if l.dataType == FloatType && r.dataType == DoubleType =>
         NaNvl(Cast(l, DoubleType), r)
       case NaNvl(l, r) if r.dataType == NullType => NaNvl(l, Cast(r, l.dataType))
+
+      case a @ ApproximatePercentile(child, pe @ NumericType(), _, _, _) =>
+        val newChild = if (TypeCollection(StringType, NullType).acceptsType(child.dataType)) {
```

Comment:
nit: I think `child.dataType == StringType || child.dataType == NullType` is clearer